### PR TITLE
Add Question matching baseline

### PR DESCRIPTION
**Pull request by tianxin1860**
### PR types

New features
### PR changes

Models
### Description

1. add question matching baseline
2. Implement R-drop as paddlenlp api

### CODE REVIEW

File: examples/text_matching/question_matching/README.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+│   └── vocab.txt
+└── ...
+```
+
+**NOTE:**
+* 如需恢复模型训练，则可以设置`init_from_ckpt`， 如`init_from_ckpt=checkpoints/model_100/model_state.pdparams`。
+
+
+### 开始预测
+训练完成后，在指定的 checkpoints 路径下会自动存储在验证集评估指标最高的模型(model_step 越大验证集指标越高)，运行如下命令开始生成预测结果:
```
Comment (ZeyuChen) [inline examples/text_matching/question_matching/README.md]: `(model_step 越大验证集指标越高)`
这句话一定正确吗？感觉会不会有点不够严谨，是否去掉？

File: examples/text_matching/question_matching/README.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
+# 短文本 query 匹配竞赛基线
+
+我们基于预训练模型 ERNIE-Gram 在短文本 query 匹配竞赛赛题上建立了 Baseline 方案和评测结果.
```
Comment (ZeyuChen) [inline examples/text_matching/question_matching/README.md]: 是否对其中R-Drop等策略加以说明并添加相关Reference

File: examples/text_matching/question_matching/model.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                token_type_ids=None,
+                position_ids=None,
+                attention_mask=None,
+                do_evaluate=False):
+
+        _, cls_embedding1 = self.ptm(input_ids, token_type_ids, position_ids,
+                                    attention_mask)
+        cls_embedding1 = self.dropout(cls_embedding1)
+        logits1 = self.classifier(cls_embedding1)
+
```
Comment (ZeyuChen) [inline examples/text_matching/question_matching/model.py]: 是否可以补充下策略名？
建议补充下参考的策略依据会显得更加专业

File: examples/text_matching/question_matching/predict.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+args = parser.parse_args()
+# yapf: enable
+
+
+def predict(model, data_loader):
+    """
+    Predicts the data labels.
+
+    Args:
+        model (obj:`SemanticIndexBase`): A model to extract text embedding or calculate similarity of text pair.
```
Comment (ZeyuChen) [inline examples/text_matching/question_matching/predict.py]: obj的类型似乎不太正确

File: examples/text_matching/question_matching/model.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    def forward(self,
+                input_ids,
+                token_type_ids=None,
+                position_ids=None,
+                attention_mask=None,
+                do_evaluate=False):
+
+        _, cls_embedding1 = self.ptm(input_ids, token_type_ids, position_ids,
+                                    attention_mask)
+        cls_embedding1 = self.dropout(cls_embedding1)
```
Comment (ZeyuChen) [inline examples/text_matching/question_matching/model.py:42]: 这里的drop out与策略有关吗？

File: examples/text_matching/question_matching/README.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
+# 短文本 query 匹配竞赛基线
```
Comment (ZeyuChen) [inline examples/text_matching/question_matching/README.md]: query 匹配这个标题不太清晰，与question matching看上去像两个任务。是否补充下相关比赛的信息

File: examples/text_matching/question_matching/model.py (hunk 3 above)
Comment (tianxin1860) [reply examples/text_matching/question_matching/model.py]: 已补充

File: examples/text_matching/question_matching/README.md (hunk 2 above)
Comment (tianxin1860) [reply examples/text_matching/question_matching/README.md]: 已补充

File: examples/text_matching/question_matching/README.md (hunk 1 above)
Comment (tianxin1860) [reply examples/text_matching/question_matching/README.md]: 已修正，模型训练部分加了 save checkpoint 的详细说明

File: examples/text_matching/question_matching/predict.py (hunk 4 above)
Comment (tianxin1860) [reply examples/text_matching/question_matching/predict.py]: 已修正，thanks~

File: examples/text_matching/question_matching/model.py (hunk 5 above)
Comment (tianxin1860) [reply examples/text_matching/question_matching/model.py:42]: 这个 dropout 和策略无关

File: examples/text_matching/question_matching/README.md (hunk 6 above)
Comment (tianxin1860) [reply examples/text_matching/question_matching/README.md]: 已修正，待陈艳比赛文案定稿之后，README 部分会再对齐更新一次。

File: examples/text_matching/question_matching/model.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        logits1 = self.classifier(cls_embedding1)
+        
+        # For more information about R-drop please refer to this paper: https://arxiv.org/abs/2106.14448
+        # Original implementation please refer to this code: https://github.com/dropreg/R-Drop
+        if self.rdrop_coef > 0 and not do_evaluate:
+            _, cls_embedding2 = self.ptm(input_ids, token_type_ids, position_ids,
+                                    attention_mask)
+            cls_embedding2 = self.dropout(cls_embedding2)
+            logits2 = self.classifier(cls_embedding2)
+            kl_loss = ppnlp.layers.rdrop_loss(logits1, logits2)
```
Comment (tianxin1860) [inline examples/text_matching/question_matching/model.py]: R-Drop 作为一个 api 添加在了 ppnlp.layers  package 下的 rdrop.py module 中

**Review by tianxin1860 — COMMENTED** [review]
add R-Drop description

File: examples/text_matching/question_matching/model.py (hunk 7 above)
Comment (ZeyuChen) [reply examples/text_matching/question_matching/model.py]: OK

File: paddlenlp/layers/rdrop.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (ZeyuChen) [inline paddlenlp/layers/rdrop.py]: 你觉得创建多一个loss/rdrop.py 放在paddlenlp顶层上是否会更好
from paddlenlp.loss import rdrop_loss

**Review by ZeyuChen — COMMENTED** [review]
探讨下是否将rdrop改成paddlenlp/loss/rdrop.py
这样后续可以拓展更多对比学习的loss

File: paddlenlp/layers/rdrop.py (hunk 8 above)
Comment (tianxin1860) [reply paddlenlp/layers/rdrop.py]: 已新增 paddlenlp.losses 模块, 后续 loss 相关会逐步补充到 paddlenlp.losses 模块下

**Review by ZeyuChen — APPROVED** [review]
LGTM
